src/vm/js/perl6-runtime/runtime.js

```diff
@@ -0,0 +1,361 @@
+
+      if (isCoercive) {
+        targetType = how.target_type(ctx, null, how, rtype);
+        rtype = how.constraint_type(ctx, null, how, rtype);
+      }
+
+      const decontValue = rv.$$decont(ctx);
+      if (decontValue.$$istype(ctx, rtype) === 0 && decontValue.$$istype(ctx, bypassType) === 0) {
+        const thrower = getThrower("X::TypeCheck::Return");
+        if (thrower === Null) {
+            ctx.die("Type check failed for return value");
+        } else {
+            thrower.$$call(ctx, null, decontValue, rtype);
+        }
+      }
+
+      if (targetType !== undefined && targetType !== rtype) {
+        const targetTypeName = targetType._STable.HOW.name(
+          ctx, null, targetType._STable.HOW, targetType).$$getStr();
+        if (rv.$$can(ctx, targetTypeName)) {
+          return rv[targetTypeName](ctx, null, rv);
+        } else {
+          const rtypeName = rtype._STable.HOW.name(ctx, null, rtype._STable.HOW, rtype).$$getStr();
+          throw new nqp.NQPException(
+            `Unable to coerce the return value from ${rtypeName} to ${targetTypeName} ;` +
+              `no coercion method defined`);
+        }
+      }
+    }
+
+    return rv;
+  };
+
+  op.p6setfirstflag = function(codeObj) {
+    firstPhaserCodeBlock = codeObj.$$getattr(Code, "$!do");
+    return codeObj;
+  };
+
+  op.p6takefirstflag = function(ctx) {
+    const matches = firstPhaserCodeBlock === ctx.codeRef();
+    firstPhaserCodeBlock = Null;
+    return matches ? 1 : 0;
+  };
+
+  const prePhaserFrames = [];
+
+  op.p6setpre = function(ctx) {
+    prePhaserFrames.push(ctx);
+    return Null;
+  };
+
+  op.p6clearpre = function(ctx) {
+    const index = prePhaserFrames.indexOf(ctx);
+    if (index !== -1) {
+      prePhaserFrames.splice(index, 1);
+    }
+    return Null;
+  };
+
+  op.p6inpre = function(ctx) {
+    const index = prePhaserFrames.indexOf(ctx.$$caller);
+    if (index !== -1) {
+      prePhaserFrames.splice(index, 1);
+      return 1;
+    } else {
+      return 0;
+    }
+  };
+
+  op.p6captureouters2 = function(ctx, capList, target) {
+    const cf = target.outerCtx;
+
+    if (cf === Null) {
+      return capList;
+    }
+
+    const elems = capList.$$elems();
+
+    for (let i = 0; i < elems; i++) {
+        const codeObj = capList.$$atpos(i);
+
+        const closure = codeObj.$$getattr(Code, "$!do");
+
+        const ctxToDiddle = closure.outerCtx;
+        if (ctxToDiddle) {
+          ctxToDiddle.$$outer = cf;
+        } else {
+          console.log("can't diddle", closure);
+        }
+    }
+
+    return capList;
+  };
+
+  op.p6capturelex = function(ctx, codeObj) {
+    const closure = codeObj.$$getattr(Code, "$!do");
+    const wantedStaticInfo = closure.staticCode.outerCodeRef;
+
+    if (ctx.codeRef().staticCode === wantedStaticInfo) {
+      closure.outerCtx = ctx;
+      closure.capture(closure.staticCode.freshBlock());
+    } else {
+      /* HACK - workaround for rakudo bugs */
```

Comment:
Is this still needed?

Reply:
I think so, rakudo still sometimes emits broken closures